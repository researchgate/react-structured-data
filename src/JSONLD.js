import React from "react";
import PropTypes from "prop-types";
import GenericNodeCollection from "./core/GenericNodeCollection";

export const JSONLD = props => {
  let json = null;
  if (props.children) {
    const ChildClass = props.children.type;
    const { children, type, ...schema } = props.children.props;

    const childClassInstance = new ChildClass(props.children.props);
    const firstChild = childClassInstance.getJSON(schema);

    json =
      childClassInstance instanceof GenericNodeCollection
        ? Object.assign(
            { "@context": "https://schema.org/" },
            { [type]: firstChild }
          )
        : Object.assign({ "@context": "https://schema.org/" }, firstChild);
  }

  return props.dangerouslyExposeHtml ? (
    <script
      type="application/ld+json"
      dangerouslySetInnerHTML={{ __html: JSON.stringify(json) }}
    />
  ) : (
    <script type="application/ld+json">{JSON.stringify(json)}</script>
  );
};

JSONLD.propTypes = {
  additionalType: PropTypes.object,
  dangerouslyExposeHtml: PropTypes.bool,
  children: PropTypes.node
};

JSONLD.defaultProps = {
  dangerouslyExposeHtml: false
};

export default JSONLD;
