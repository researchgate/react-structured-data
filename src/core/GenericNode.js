import JSONLDAbstractNode from "../JSONLDAbstractNode";

class GenericNode extends JSONLDAbstractNode {
  getJSON({ type, jsonldtype, ...schema }) {
    const parseChildren = super.parseChildren();
    const details = {
      "@type": jsonldtype || this.getSchemaOrgType(),
      ...schema
    };
    Object.keys(details).forEach(key => {
      if (details[key] === null) delete details[key];
    });
    return type
      ? Object.assign({ [type]: details }, ...parseChildren)
      : Object.assign(details, ...parseChildren);
  }
}

export default GenericNode;
