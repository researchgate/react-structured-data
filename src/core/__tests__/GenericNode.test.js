import React from "react";
import renderer from "react-test-renderer";
import JSONLD from "../../JSONLD";
import GenericNode from "../GenericNode";

describe("GenerticNode", () => {
  it("renders question to QAPage mainEntity as Question type", () => {
    expect(
      renderer.create(
        <JSONLD>
          <GenericNode
            type="review"
            jsonldtype="Review"
            name="It is awesome"
            reviewBody="This is great!"
          >
            <GenericNode
              type="itemReviewed"
              jsonldtype="Product"
              id="product-x"
            />
            <GenericNode type="author" jsonldtype="Person" name="Cool Carl" />
            <GenericNode
              type="locationCreated"
              jsonldtype="AdministrativeArea"
              name="Chicago, IL"
            />
          </GenericNode>
        </JSONLD>
      )
    ).toMatchSnapshot();
  });
});
