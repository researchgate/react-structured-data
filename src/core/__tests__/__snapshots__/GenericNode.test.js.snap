// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenerticNode filters out null and false child nodes 1`] = `
<script
  type="application/ld+json"
>
  {"@context":"https://schema.org/","@type":"Review","name":"It is awesome","reviewBody":"This is great!","itemReviewed":{"@type":"Product","@id":"product-x"},"locationCreated":{"@type":"AdministrativeArea","name":"Chicago, IL"}}
</script>
`;

exports[`GenerticNode filters out properties that have null values 1`] = `
<script
  type="application/ld+json"
>
  {"@context":"https://schema.org/","@type":"Review","name":"It is awesome"}
</script>
`;

exports[`GenerticNode renders question to QAPage mainEntity as Question type 1`] = `
<script
  type="application/ld+json"
>
  {"@context":"https://schema.org/","@type":"Review","name":"It is awesome","reviewBody":"This is great!","itemReviewed":{"@type":"Product","@id":"product-x"},"author":{"@type":"Person","name":"Cool Carl"},"locationCreated":{"@type":"AdministrativeArea","name":"Chicago, IL"}}
</script>
`;
